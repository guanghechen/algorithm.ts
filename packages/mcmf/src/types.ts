/**
 * 残量网络中的一条弧
 * An arc in the residual network.
 */
export interface McmfEdge {
  /**
   * The starting node of the arc.
   */
  from: number
  /**
   * The ending node of the arc.
   */
  to: number
  /**
   * Capacity of the arc.
   */
  cap: number
  /**
   * Flow on the arc.
   */
  flow: number
  /**
   * Price of every unit flow.
   */
  cost: number
}

export interface McmfContext {
  /**
   *
   */
  edgeTot: number
  /**
   *
   */
  dist: number[]
  /**
   *
   */
  edges: McmfEdge[]
  /**
   *
   */
  G: number[][]
}

export interface Mcmf {
  /**
   * Initialize the mcmf algorithm.
   * @param source  the source node
   * @param target  the target node
   * @param n       the number of nodes
   * @param m       the number of edges
   */
  init(source: number, target: number, n: number, m: number): void

  /**
   * Add an edge into the residual network.
   * @param from    the starting node
   * @param to      the ending node
   * @param cap     capacity
   * @param cost    unit price of flow
   */
  addEdge(from: number, to: number, cap: number, cost: number): void

  /**
   * Calculate the min cost and max flow of the residual network.
   * @returns [minCost, maxFlow]
   */
  minCostMaxFlow(): [number, number]

  /**
   * A hook to support access to mcmf internal variables: it accepts a callback
   * function, and when it is executed, part of the internal variables of mcmf
   * will be passed into the callback function as parameters.
   * @param fn
   */
  solve(fn: (context: McmfContext) => void): void
}
