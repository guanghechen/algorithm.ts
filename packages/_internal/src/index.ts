export * from './constant'
export * from './types'
