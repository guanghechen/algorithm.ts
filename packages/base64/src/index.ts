import { Base64 } from './base64'

export * from './base64'

const base64 = new Base64()

export const encode = base64.encode.bind(base64)
export const decode = base64.decode.bind(base64)
export const validate = base64.validate.bind(base64)
export default { encode, decode, validate }
