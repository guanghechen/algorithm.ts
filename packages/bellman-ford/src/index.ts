import { createCircularQueue } from '@algorithm.ts/circular-queue'

export interface IEdge {
  /**
   * The other end of the edge.
   */
  to: number
  /**
   * The cost of walking along this side.
   */
  cost: number
}

export interface IGraph {
  /**
   * The number of nodes in the graph. (0-index)
   */
  N: number
  /**
   * The source node. (0-index)
   */
  source: number
  /**
   * Graph edges.
   */
  edges: ReadonlyArray<IEdge>
  /**
   * Adjacency list. G[i] represent the index list of the  edges start from node i.
   */
  G: ReadonlyArray<ReadonlyArray<number>>
  /**
   * An array recording the shortest distance to the source point.
   */
  dist: number[]
}

export interface IOptions {
  /**
   * A big number, representing the unreachable cost.
   */
  INF?: number
  /**
   * Used to check if an element is already in the queue.
   */
  inq?: boolean[]
  /**
   * Record the number of times an element is enqueued,
   * used to check whether there is a negative cycle.
   */
  inqTimes?: number[]
}

const ZERO = 0
const DEFAULT_INF = Math.floor(Number.MAX_SAFE_INTEGER / 2)
const Q = createCircularQueue<number>()

/**
 * The bellman-ford algorithm, optimized with queue.
 *
 * @param graph
 * @param options
 */
export function bellmanFord(graph: IGraph, options: IOptions = {}): boolean {
  const { N, source, edges, G, dist } = graph
  const { INF = DEFAULT_INF, inq = [], inqTimes = [] } = options

  if (dist.length < N) dist.length = N
  if (inq.length < N) inq.length = N
  if (inqTimes.length < N) inqTimes.length = N

  dist.fill(INF, 0, N)
  inq.fill(false, 0, N)
  inqTimes.fill(0, 0, N)

  Q.init(N + 1)
  Q.enqueue(source)
  dist[source] = ZERO

  while (Q.size() > 0) {
    const o: number = Q.dequeue()!
    inq[o] = false

    for (const idx of G[o]) {
      const edge: IEdge = edges[idx]
      if (dist[edge.to] > dist[o] + edge.cost) {
        dist[edge.to] = dist[o] + edge.cost
        if (!inq[edge.to]) {
          Q.enqueue(edge.to)
          inq[edge.to] = true

          // eslint-disable-next-line no-plusplus
          if (++inqTimes[edge.to] > N) return false
        }
      }
    }
  }
  return true
}

export default bellmanFord
