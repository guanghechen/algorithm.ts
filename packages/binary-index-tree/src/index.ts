export * from './tree1'
export * from './tree2'
export * from './types'
export * from './util'
