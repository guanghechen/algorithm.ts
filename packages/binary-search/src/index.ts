export * from './binary-search'
export * from './lower-bound'
export * from './upper-bound'
export * from './types'
