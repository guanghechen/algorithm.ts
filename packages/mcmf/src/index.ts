import type { ICircularQueue } from '@algorithm.ts/circular-queue'
import { createCircularQueue } from '@algorithm.ts/circular-queue'

/**
 * An arc (or edge) in the residual network.
 */
export interface IEdge {
  /**
   * The starting node of the arc.
   */
  from: number
  /**
   * The ending node of the arc.
   */
  to: number
  /**
   * Capacity of the arc.
   */
  cap: number
  /**
   * Flow on the arc.
   */
  flow: number
  /**
   * Price of every unit flow.
   */
  cost: number
}

export interface IContext {
  /**
   * The number of valid edges.
   *
   * In order to save space, there is not recreate a new array at each `init()` called, so this
   * field is used to record the actual effective array length.
   */
  edgeTot: number
  /**
   * Graph edges.
   */
  edges: ReadonlyArray<IEdge>
  /**
   * Adjacency list. G[i] represent the index list of the  edges start from node i.
   */
  G: ReadonlyArray<ReadonlyArray<number>>
}

export interface IMcmf {
  /**
   * Initialize the mcmf algorithm.
   * @param source  the source node
   * @param target  the target node
   * @param n       the number of nodes
   */
  init(source: number, target: number, n: number): void

  /**
   * Add an edge into the residual network.
   * @param from    the starting node
   * @param to      the ending node
   * @param cap     capacity
   * @param cost    unit price of flow
   */
  addEdge(from: number, to: number, cap: number, cost: number): void

  /**
   * Calculate the min cost and max flow of the residual network.
   * @returns [minCost, maxFlow]
   */
  minCostMaxFlow(): [number, number]

  /**
   * A hook to support access to mcmf internal variables: it accepts a callback
   * function, and when it is executed, part of the internal variables of mcmf
   * will be passed into the callback function as parameters.
   * @param fn
   */
  solve(fn: (context: IContext) => void): void
}

export interface IOptions {
  /**
   * A big number, representing the unreachable cost.
   */
  INF?: number
}

const DEFAULT_INF = Math.floor(Number.MAX_SAFE_INTEGER / 2)

export function createMcmf(options: IOptions = {}): IMcmf {
  let _source: number // The source point in a network flow
  let _target: number // The sink in a network flow
  let _n: number // The number of nodes in a network flow
  let _mincost: number
  let _maxflow: number
  let _edgeTot: number
  const _INF: number = options.INF ?? DEFAULT_INF
  const _inq: boolean[] = [] // Whether if the i-th node is in the queue.
  const _dist: number[] = [] // The distance from the source node to the i-th node.
  const _path: number[] = [] // An edge in an augmented path.
  const _edges: IEdge[] = []
  const _G: number[][] = []
  const _Q: ICircularQueue<number> = createCircularQueue()
  return { init, addEdge, minCostMaxFlow, solve }

  function init(source: number, target: number, n: number): void {
    _source = source
    _target = target
    _n = n
    _mincost = 0
    _maxflow = 0

    // Resize arrays.
    if (_inq.length < _n) {
      const _start = _inq.length
      _inq.length = _n
      _inq.fill(false, _start, _n)
    }
    if (_dist.length < _n) _dist.length = _n
    if (_path.length < _n) _path.length = _n
    if (_G.length < _n) _G.length = _n

    _edgeTot = 0
    _Q.init(_n + 1)
    for (let i = 0; i < _n; ++i) _G[i] = []
  }

  function addEdge(from: number, to: number, cap: number, cost: number): void {
    _G[from].push(_edgeTot)
    // eslint-disable-next-line no-plusplus
    _edges[_edgeTot++] = { from, to, cap, flow: 0, cost }

    _G[to].push(_edgeTot)
    // eslint-disable-next-line no-plusplus
    _edges[_edgeTot++] = { from: to, to: from, cap: 0, flow: 0, cost: -cost }
  }

  function minCostMaxFlow(): [number, number] {
    while (bellmanFord()) {
      let mif = _INF
      for (let o = _target; o !== _source; ) {
        const e = _edges[_path[o]]
        const remainCap = e.cap - e.flow
        if (mif > remainCap) mif = remainCap
        o = e.from
      }
      for (let o = _target; o !== _source; ) {
        const x = _path[o]
        _edges[x].flow += mif
        _edges[x ^ 1].flow -= mif
        o = _edges[x].from
      }
      _maxflow += mif
      _mincost += mif * _dist[_target]
    }
    return [_mincost, _maxflow]
  }

  function solve(fn: (context: IContext) => void): void {
    const context: IContext = {
      edgeTot: _edgeTot,
      edges: _edges,
      G: _G,
    }
    fn(context)
  }

  /**
   * Negative loops should not appear in the residual network,
   * so there is no need to check if there are negative loops.
   */
  function bellmanFord(): boolean {
    // Initialize the dist array.
    _dist.fill(_INF, 0, _n)

    _Q.enqueue(_source)
    _dist[_source] = 0
    _inq[_source] = true
    while (_Q.size() > 0) {
      const o = _Q.dequeue()!
      for (let i = 0, g = _G[o]; i < g.length; ++i) {
        const x = g[i]
        const e = _edges[x]
        if (e.cap === e.flow) continue

        const candidateDist = _dist[o] + e.cost
        if (_dist[e.to] > candidateDist) {
          _dist[e.to] = candidateDist
          _path[e.to] = x

          if (_inq[e.to]) continue
          _inq[e.to] = true
          _Q.enqueue(e.to)
        }
      }
      _inq[o] = false
    }

    return _dist[_target] !== _INF
  }
}
