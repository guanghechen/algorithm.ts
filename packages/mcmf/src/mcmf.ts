import type { ICircularQueue } from '@algorithm.ts/circular-queue'
import { createCircularQueue } from '@algorithm.ts/circular-queue'
import type { Mcmf, McmfContext, McmfEdge } from './types'

export function createMcmf(): Mcmf {
  let _source: number // The source point in a network flow
  let _target: number // The sink in a network flow
  let _n: number // The number of nodes in a network flow
  let _mincost: number
  let _maxflow: number
  let _edgeTot: number
  const _inq: boolean[] = [] // Whether if the i-th node is in the queue.
  const _dist: number[] = [] // The distance from the source node to the i-th node.
  const _path: number[] = [] // An edge in an augmented path.
  const _edges: McmfEdge[] = []
  const _G: number[][] = []
  const _Q: ICircularQueue<number> = createCircularQueue()
  return { init, addEdge, minCostMaxFlow, solve }

  function init(source: number, target: number, n: number): void {
    _source = source
    _target = target
    _n = n
    _mincost = 0
    _maxflow = 0

    // Resize arrays.
    if (_inq.length < _n) {
      const _start = _inq.length
      _inq.length = _n
      _inq.fill(false, _start, _n)
    }
    if (_dist.length < _n) _dist.length = _n
    if (_path.length < _n) _path.length = _n
    if (_G.length < _n) _G.length = _n

    _edgeTot = 0
    _Q.init(_n + 1)
    for (let i = 0; i < _n; ++i) _G[i] = []
  }

  function addEdge(from: number, to: number, cap: number, cost: number): void {
    _G[from].push(_edgeTot)
    // eslint-disable-next-line no-plusplus
    _edges[_edgeTot++] = { from, to, cap, flow: 0, cost }

    _G[to].push(_edgeTot)
    // eslint-disable-next-line no-plusplus
    _edges[_edgeTot++] = { from: to, to: from, cap: 0, flow: 0, cost: -cost }
  }

  function minCostMaxFlow(): [number, number] {
    while (spfa()) {
      let mif = Number.MAX_SAFE_INTEGER
      for (let o = _target; o !== _source; ) {
        const e = _edges[_path[o]]
        const remainCap = e.cap - e.flow
        if (mif > remainCap) mif = remainCap
        o = e.from
      }
      for (let o = _target; o !== _source; ) {
        const x = _path[o]
        _edges[x].flow += mif
        _edges[x ^ 1].flow -= mif
        o = _edges[x].from
      }
      _maxflow += mif
      _mincost += mif * _dist[_target]
    }
    return [_mincost, _maxflow]
  }

  function solve(fn: (context: McmfContext) => void): void {
    const context: McmfContext = {
      edgeTot: _edgeTot,
      dist: _dist,
      edges: _edges,
      G: _G,
    }
    fn(context)
  }

  function spfa(): boolean {
    // Initialize the dist array.
    _dist.fill(-1, 0, _n)

    _Q.enqueue(_source)
    _dist[_source] = 0
    _inq[_source] = true
    while (_Q.size() > 0) {
      const o = _Q.dequeue()!
      for (let i = 0, g = _G[o]; i < g.length; ++i) {
        const x = g[i]
        const e = _edges[x]
        if (e.cap === e.flow) continue

        const candidateDist = _dist[o] + e.cost
        if (_dist[e.to] === -1 || _dist[e.to] > candidateDist) {
          _dist[e.to] = candidateDist
          _path[e.to] = x

          if (_inq[e.to]) continue
          _inq[e.to] = true
          _Q.enqueue(e.to)
        }
      }
      _inq[o] = false
    }

    return _dist[_target] !== -1
  }
}
