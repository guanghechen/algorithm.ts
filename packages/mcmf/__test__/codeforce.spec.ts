import fs from 'fs-extra'
import { createReader, loadFixtures, locateFixtures } from 'jest.setup'
import path from 'path'
import { createMcmf } from '../src'

describe('codeforces', function () {
  describe('contest 1082', function () {
    test('g', function () {
      const mcmf = createMcmf()
      function solve(content: string): number {
        const io = createReader(content)
        const [n, m] = io.readIntegersOfLine()

        const source = 0
        const target: number = n + m + 1
        mcmf.init(source, target, n + m + 2, n + m * 3)

        const nodes: number[] = io.readIntegersOfLine()
        for (let i = 0; i < n; ++i) {
          const weight: number = nodes[i]
          mcmf.addEdge(i + 1, target, weight, 1)
        }

        let answer = 0
        for (let i = 1; i <= m; ++i) {
          const [u, v, weight] = io.readIntegersOfLine()
          const x = n + i
          answer += weight
          mcmf.addEdge(source, x, weight, 1)
          mcmf.addEdge(x, u, Number.MAX_SAFE_INTEGER, 1)
          mcmf.addEdge(x, v, Number.MAX_SAFE_INTEGER, 1)
        }

        const [mincost, maxflow] = mcmf.minCostMaxFlow()
        answer -= maxflow
        return answer
      }

      const caseDir: string = locateFixtures('codeforces/1082/g')
      const filenames: string[] = fs.readdirSync(caseDir).sort()
      for (const filename of filenames) {
        if (filename.endsWith('.in')) {
          const inputFilepath = path.join(caseDir, filename)
          const content = loadFixtures(inputFilepath)
          const output: number = solve(content)

          const answerFilepath = inputFilepath.replace(/\.in$/, '.out')
          const answer = Number(loadFixtures(answerFilepath))
          // eslint-disable-next-line jest/no-conditional-expect
          expect(output).toEqual(answer)
        }
      }

      mcmf.solve(context => expect(Object.keys(context)).toMatchSnapshot())
    })
  })

  describe('contest 277', function () {
    test('e', function () {
      const mcmf = createMcmf()

      const caseDir: string = locateFixtures('codeforces/0277/e')
      const filenames: string[] = fs.readdirSync(caseDir).sort()
      for (const filename of filenames) {
        if (filename.endsWith('.in')) {
          const inputFilepath = path.join(caseDir, filename)
          const content = loadFixtures(inputFilepath)
          const output: number = solve(content)
          const answerFilepath = inputFilepath.replace(/\.in$/, '.out')
          const answer = Number(loadFixtures(answerFilepath))
          // eslint-disable-next-line jest/no-conditional-expect
          expect(Math.abs(output - answer)).toBeLessThanOrEqual(1e-6)
        }
      }
      mcmf.solve(context => expect(Object.keys(context)).toMatchSnapshot())

      function solve(content: string): number {
        const io = createReader(content)

        const [N] = io.readIntegersOfLine()

        const vertexes: Vertex[] = new Array(N)
        for (let i = 0; i < N; ++i) {
          const [x, y] = io.readIntegersOfLine()
          vertexes[i] = { x, y }
        }
        vertexes.sort((p, q) => {
          if (p.y === q.y) return p.x - q.x
          return q.y - p.y
        })

        const source = 0
        const target: number = N * 2 + 1
        mcmf.init(source, target, N * 2 + 2, N * N + 2 * N)

        for (let i = 0; i < N; ++i) {
          mcmf.addEdge(source, i + 1, 2, 0)
          mcmf.addEdge(N + i + 1, target, 1, 0)
          for (let j = i + 1; j < N; ++j) {
            if (vertexes[i].y === vertexes[j].y) continue
            mcmf.addEdge(i + 1, N + j + 1, 1, dist(vertexes[i], vertexes[j]))
          }
        }

        const [mincost, maxflow] = mcmf.minCostMaxFlow()
        const answer = maxflow === N - 1 ? mincost : -1
        return answer
      }

      function dist(p: Vertex, q: Vertex): number {
        const d = (p.x - q.x) * (p.x - q.x) + (p.y - q.y) * (p.y - q.y)
        return Math.sqrt(d)
      }

      interface Vertex {
        x: number
        y: number
      }
    })
  })
})
