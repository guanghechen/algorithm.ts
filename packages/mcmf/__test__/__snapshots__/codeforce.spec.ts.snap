// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codeforces contest 277 e 1`] = `
Array [
  "edgeTot",
  "dist",
  "edges",
  "G",
]
`;

exports[`codeforces contest 1082 g 1`] = `
Array [
  "edgeTot",
  "dist",
  "edges",
  "G",
]
`;
